web/interceptor/helpers.go

```diff
@@ -14,45 +14,60 @@ import (
 // getAuthenticatedContext returns a new context with the user ID attached to it.
 // If the context does not contain a valid session cookie, it returns an error.
 func getAuthenticatedContext(ctx context.Context, logger *zap.SugaredLogger, tm *auth.TokenManager) (context.Context, error) {
+	claims, err := tm.GetClaims(ctx)
+	if err != nil {
+		logger.Errorf("Failed to extract claims from JWT: %v", err)
+		return nil, ErrInvalidAuthCookie
+	}
 	meta, ok := metadata.FromIncomingContext(ctx)
 	if !ok {
 		logger.Error("Failed to extract metadata")
 		return nil, ErrContextMetadata
 	}
-	token, err := extractToken(meta)
-	if err != nil {
-		logger.Errorf("Failed to extract token: %v", err)
-		return nil, err
-	}
-	claims, err := tm.GetClaims(token)
-	if err != nil {
-		logger.Errorf("Failed to extract claims from JWT: %v", err)
-		return nil, ErrInvalidAuthCookie
-	}
 	if tm.UpdateRequired(claims) {
-		logger.Debug("Updating token for user ", claims.UserID)
-		updatedToken, err := tm.NewAuthCookie(claims.UserID)
+		logger.Debug("Updating cookie for user ", claims.UserID)
```

Comment:
Suggest that this if-clause adjusted like this:
```go
logger.Debug("Updating cookie for user ", claims.UserID)
meta, err = tm.UpdateCookie(ctx, claims)
if err != nil {
	logger.Errorf("Failed to update cookie: %v", err)
	return nil, ErrInvalidAuthCookie
}
```
And the corresponding method in `jwt_update.go` could look something like this:
```go
func (tm *TokenManager) UpdateCookie(ctx context.Context, claims *Claims) (metadata.MD, error) {
	updatedCookie, err := tm.NewAuthCookie(claims.UserID)
	if err != nil {
		return nil, fmt.Errorf("failed to update cookie for user %d: %w", claims.UserID, err)
	}
	if err := tm.Remove(claims.UserID); err != nil {
		return nil, err
	}
	if err := grpc.SendHeader(ctx, metadata.Pairs(SetCookie, updatedCookie.String())); err != nil {
		return nil, fmt.Errorf("failed to set grpc header: %w", err)
	}
	return metadata.New(map[string]string{
		Cookie: TokenString(updatedCookie),
	}), nil
}
```
The only thing I'm not sure about is whether or not to pass `ctx` required for `grpc.SendHeader` or to just return the `updatedCookie`, and leave the metadata stuff in the `getAuthenticatedContext` func, like this:
```go
func (tm *TokenManager) UpdateCookie(ctx context.Context, claims *Claims) (*http.Cookie, error) {
	updatedCookie, err := tm.NewAuthCookie(claims.UserID)
	if err != nil {
		return nil, fmt.Errorf("failed to update cookie for user %d: %w", claims.UserID, err)
	}
	if err := tm.Remove(claims.UserID); err != nil {
		return nil, err
	}
	return updatedCookie, nil
}
```
(also should the `Remove` be done before the `NewAuthCookie`?)

Reply:
But what if there is an error while updating a cookie? Then `Remove` would remove user from the token update list, and the user's JWT would never get updated before it expires.

Reply:
Okay. I guess it is just me that didn't recall the details of the Remove func.